gflownet/envs/base.py

```diff
@@ -242,8 +282,27 @@ def sample_actions(
             logits = policy_outputs
             logits /= temperature_logits
         if mask_invalid_actions is not None:
+            assert not torch.all(
+                mask_invalid_actions
+            ), """
+            All actions in the mask are invalid.
+            """
             logits[mask_invalid_actions] = -loginf
-        action_indices = Categorical(logits=logits).sample()
+        else:
+            mask_invalid_actions = torch.zeros(
+                policy_outputs.shape, dtype=torch.bool, device=device
+            )
+        # Make sure that a valid action is sampled, otherwise throw an error.
+        for _ in range(max_sampling_attempts):
+            action_indices = Categorical(logits=logits).sample()
+            if not torch.any(mask_invalid_actions[ns_range, action_indices]):
+                break
+        else:
```

Comment:
Am I seeing this misindented on Github for some reason or is it a bug?

![SCR-20230626-mqby](https://github.com/alexhernandezgarcia/gflownet/assets/9283470/18754efa-017d-4a12-a6c1-54f87564b4f1)

Reply:
```
        for _ in range(max_sampling_attempts):
            action_indices = Categorical(logits=logits).sample()
            if not torch.any(mask_invalid_actions[ns_range, action_indices]):
                logprobs = self.logsoftmax(logits)[ns_range, action_indices]
                # Build actions
                actions = [self.action_space[idx] for idx in action_indices]
                return actions, logprobs
        raise ValueError(
            f"No valid action could be sampled after {max_sampling_attempts} attempts."
        )
```

(or use `dedent` with `from textwrap import dedent` at the top of the file)

Reply:
Is your comment about the `else` part? If so, it is correct as you see it. It's a `for ... else` loop. If the `break` clause is not reached during the for loop, then the code gets into the `else` part. So this piece of code attempts to `max_sampling_attempts` times to sample actions that are not invalid, otherwise raises an error.

```python
import numpy as np
max_n = 10
threshold = 0.9
for _ in range(max_n):
    if np.random.uniform() > threshold:
        break
else:
    print(f"No random number in [0, 1] larger than {threshold} in {max_n} uniform samples")
```

Reply:
Amazing I didn't know about that!!

Researching this, I also learned about `try/except/*else*/finally` clauses

```python
try:
    a()
except Exception:
    # executed if a() throws an exception
    b()
else:
	# executed if a() does not throw an exception BUT c() won't be excepted
    c()
finally:
    # always executed
    d()
```

Reply:
I never thought I would teach you any Python! :P 

I've also just learnt about `try/except/*else*/finally`